src/stream/src/executor/top_n/group_top_n_appendonly.rs

```diff
@@ -0,0 +1,214 @@
+    InnerAppendOnlyGroupTopNExecutorNew<K, S, WITH_TIES>
+{
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        input_info: ExecutorInfo,
+        storage_key: Vec<OrderPair>,
+        offset_and_limit: (usize, usize),
+        order_by: Vec<OrderPair>,
+        executor_id: u64,
+        group_by: Vec<usize>,
+        state_table: StateTable<S>,
+        lru_manager: AtomicU64Ref,
+    ) -> StreamResult<Self> {
+        let ExecutorInfo {
+            pk_indices, schema, ..
+        } = input_info;
+
+        let cache_key_serde =
+            create_cache_key_serde(&storage_key, &pk_indices, &schema, &order_by, &group_by);
+        let managed_state = ManagedTopNState::<S>::new(state_table, cache_key_serde.clone());
+
+        Ok(Self {
+            info: ExecutorInfo {
+                schema,
+                pk_indices,
+                identity: format!("AppendOnlyGroupTopNExecutor {:X}", executor_id),
+            },
+            offset: offset_and_limit.0,
+            limit: offset_and_limit.1,
+            managed_state,
+            storage_key_indices: storage_key.into_iter().map(|op| op.column_idx).collect(),
+            group_by,
+            caches: GroupTopNCache::new(lru_manager),
+            cache_key_serde,
+        })
+    }
+}
+#[async_trait]
+impl<K: HashKey, S: StateStore, const WITH_TIES: bool> TopNExecutorBase
+    for InnerAppendOnlyGroupTopNExecutorNew<K, S, WITH_TIES>
+where
+    TopNCache<WITH_TIES>: AppendOnlyTopNCacheTrait,
+{
+    async fn apply_chunk(&mut self, chunk: StreamChunk) -> StreamExecutorResult<StreamChunk> {
+        let mut res_ops = Vec::with_capacity(self.limit);
+        let mut res_rows = Vec::with_capacity(self.limit);
+        let chunk = chunk.compact();
+        let keys = K::build(&self.group_by, chunk.data_chunk())?;
+
+        let data_types = self.schema().data_types();
+        let row_deserializer = RowDeserializer::new(data_types);
+
+        for ((op, row_ref), group_cache_key) in chunk.rows().zip_eq(keys.iter()) {
+            // The pk without group by
+            let pk_row = row_ref.project(&self.storage_key_indices[self.group_by.len()..]);
+            let cache_key = serialize_pk_to_cache_key(pk_row, &self.cache_key_serde);
+
+            let group_key = row_ref.project(&self.group_by);
+
+            // If 'self.caches' does not already have a cache for the current group, create a new
+            // cache for it and insert it into `self.caches`
+            if !self.caches.contains(group_cache_key) {
+                let mut topn_cache = TopNCache::new(self.offset, self.limit);
+                self.managed_state
+                    .init_topn_cache(Some(group_key), &mut topn_cache)
+                    .await?;
+                self.caches.push(group_cache_key.clone(), topn_cache);
+            }
+            let cache = self.caches.get_mut(group_cache_key).unwrap();
+
+            debug_assert_eq!(op, Op::Insert);
+            cache.insert(
```

Comment:
This make me really confused for minutes until I find the `insert` is defined in two traits. 😄